src/expressions.js

```diff
@@ -0,0 +1,45 @@
+const scopes = require('taskcluster-lib-scopes');
+const _ = require('lodash');
+
+export const splatParams = (scope, params, missing) => scope.replace(/<([^>]+)>/g, (match, param) => {
+  const value = _.at(params, param)[0];
+  if (value !== undefined) {
+    return value;
+  }
+  missing.push(match); // If any are left undefined, we can't be done yet
+  return match;
+});
+
+export const expandExpressionTemplate = (template, params, missing) => {
+  if (typeof template === 'string') {
+    return splatParams(template, params, missing);
+  } else if (_.isObject(template) && template.for && template.in && template.each) {
+    let subs = _.at(params, template.in)[0];
+    if (!subs) {
+      missing.push(template.in);
+      return;
+    }
+    return subs.map(param => splatParams(template.each.replace(`<${template.for}>`, param), params, missing));
```

Comment:
Are you sure we shouldn't do:
```js
const param2 = Object.assign({}, params);
subs.map(param => {
  param2[template.for] = param;
  return splatParams(template.each, param2);
});
```

It's is more expensive.. but probably not much...

Reply:
better name than `param2` obivously...

Reply:
Note: this is strictly more expressive...